File: DEG-main/devkits/p2p-trading-ies-wave1/beckn-signing-kit/golang/signer_test.go
package signer

import (
	"crypto/ed25519"
	"encoding/base64"
	"strings"
	"testing"
	"time"
)

// Test keys matching the config in local-p2p-bap.yaml (sandbox1).
const (
	testSubscriberID = "p2p-trading-sandbox1.com"
	testKeyID        = "76EU8aUqHouww7gawT6EibH4bseMCumyDv3sgyXSKENGk8NDcdVwmQ"
	testPrivateKey   = "Pc6dkYo5LeP0LkwvZXVRV9pcbeh8jDdtdHWymID5cjw="
	testPublicKey    = "KVYEWkQB2WwnttVMWfy7KrnqiD51ZDvi8vfCac2IwRE="
)

// Sample beckn payload (trimmed confirm request).
var samplePayload = []byte(`{
  "context": {
    "version": "2.0.0",
    "action": "confirm",
    "timestamp": "2024-10-04T10:25:00Z",
    "message_id": "msg-confirm-001",
    "transaction_id": "txn-energy-001",
    "bap_id": "bap.energy-consumer.com",
    "bap_uri": "https://bap.energy-consumer.com",
    "bpp_id": "bpp.energy-provider.com",
    "bpp_uri": "https://bpp.energy-provider.com",
    "domain": "beckn.one:deg:p2p-trading:2.0.0"
  },
  "message": {
    "order": {
      "@type": "beckn:Order",
      "beckn:orderStatus": "CREATED",
      "beckn:seller": "provider-solar-farm-001"
    }
  }
}`)

func TestSignAndVerifyRoundTrip(t *testing.T) {
	// --- Sign ---
	s, err := New(Config{
		SubscriberID:     testSubscriberID,
		UniqueKeyID:      testKeyID,
		SigningPrivateKey: testPrivateKey,
	})
	if err != nil {
		t.Fatalf("New() error: %v", err)
	}

	authHeader, err := s.SignPayload(samplePayload)
	if err != nil {
		t.Fatalf("SignPayload() error: %v", err)
	}

	// Basic structure checks
	if !strings.HasPrefix(authHeader, "Signature ") {
		t.Errorf("expected header to start with 'Signature ', got: %s", authHeader)
	}
	if !strings.Contains(authHeader, testSubscriberID+"|"+testKeyID+"|ed25519") {
		t.Errorf("keyId not found in header: %s", authHeader)
	}
	if !strings.Contains(authHeader, `algorithm="ed25519"`) {
		t.Errorf("algorithm not found in header: %s", authHeader)
	}

	t.Logf("Authorization: %s", authHeader)

	// --- Verify ---
	err = Verify(samplePayload, authHeader, testPublicKey)
	if err != nil {
		t.Fatalf("Verify() error: %v", err)
	}
}

func TestSignPayloadDetailed(t *testing.T) {
	s, err := New(Config{
		SubscriberID:     testSubscriberID,
		UniqueKeyID:      testKeyID,
		SigningPrivateKey: testPrivateKey,
	})
	if err != nil {
		t.Fatalf("New() error: %v", err)
	}

	result, err := s.SignPayloadDetailed(samplePayload)
	if err != nil {
		t.Fatalf("SignPayloadDetailed() error: %v", err)
	}

	if result.ExpiresAt-result.CreatedAt != 300 {
		t.Errorf("expected 300s expiry window, got %d", result.ExpiresAt-result.CreatedAt)
	}
	if result.Signature == "" {
		t.Error("expected non-empty signature")
	}
	if result.AuthorizationHeader == "" {
		t.Error("expected non-empty auth header")
	}
}

func TestVerifyRejectsTamperedPayload(t *testing.T) {
	s, err := New(Config{
		SubscriberID:     testSubscriberID,
		UniqueKeyID:      testKeyID,
		SigningPrivateKey: testPrivateKey,
	})
	if err != nil {
		t.Fatalf("New() error: %v", err)
	}

	authHeader, err := s.SignPayload(samplePayload)
	if err != nil {
		t.Fatalf("SignPayload() error: %v", err)
	}

	// Tamper with payload
	tampered := append([]byte{}, samplePayload...)
	tampered[10] = 'X'

	err = Verify(tampered, authHeader, testPublicKey)
	if err == nil {
		t.Fatal("expected verification to fail for tampered payload")
	}
}

func TestVerifyRejectsExpiredSignature(t *testing.T) {
	s, err := New(Config{
		SubscriberID:     testSubscriberID,
		UniqueKeyID:      testKeyID,
		SigningPrivateKey: testPrivateKey,
		ExpiryDuration:   1 * time.Second,
	})
	if err != nil {
		t.Fatalf("New() error: %v", err)
	}

	// Sign at a time in the past
	pastTime := time.Now().Add(-10 * time.Minute)
	result, err := s.signPayloadAt(samplePayload, pastTime)
	if err != nil {
		t.Fatalf("signPayloadAt() error: %v", err)
	}

	err = Verify(samplePayload, result.AuthorizationHeader, testPublicKey)
	if err == nil {
		t.Fatal("expected verification to fail for expired signature")
	}
	if !strings.Contains(err.Error(), "expired") {
		t.Errorf("expected 'expired' in error, got: %v", err)
	}
}

func TestVerifyRejectsWrongPublicKey(t *testing.T) {
	s, err := New(Config{
		SubscriberID:     testSubscriberID,
		UniqueKeyID:      testKeyID,
		SigningPrivateKey: testPrivateKey,
	})
	if err != nil {
		t.Fatalf("New() error: %v", err)
	}

	authHeader, err := s.SignPayload(samplePayload)
	if err != nil {
		t.Fatalf("SignPayload() error: %v", err)
	}

	// Use a different key pair
	_, wrongPriv, _ := ed25519.GenerateKey(nil)
	wrongPub := wrongPriv.Public().(ed25519.PublicKey)
	wrongPubB64 := base64.StdEncoding.EncodeToString(wrongPub)

	err = Verify(samplePayload, authHeader, wrongPubB64)
	if err == nil {
		t.Fatal("expected verification to fail for wrong public key")
	}
}

func TestNewValidatesConfig(t *testing.T) {
	tests := []struct {
		name string
		cfg  Config
		want string
	}{
		{"missing subscriber", Config{UniqueKeyID: "k", SigningPrivateKey: testPrivateKey}, "SubscriberID"},
		{"missing keyID", Config{SubscriberID: "s", SigningPrivateKey: testPrivateKey}, "UniqueKeyID"},
		{"missing private key", Config{SubscriberID: "s", UniqueKeyID: "k"}, "SigningPrivateKey"},
		{"bad base64", Config{SubscriberID: "s", UniqueKeyID: "k", SigningPrivateKey: "not-base64!!!"}, "invalid base64"},
		{"wrong key size", Config{SubscriberID: "s", UniqueKeyID: "k", SigningPrivateKey: base64.StdEncoding.EncodeToString([]byte("short"))}, "must be 32 bytes"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			_, err := New(tt.cfg)
			if err == nil {
				t.Fatal("expected error")
			}
			if !strings.Contains(err.Error(), tt.want) {
				t.Errorf("expected error containing %q, got: %v", tt.want, err)
			}
		})
	}
}

func TestParseKeyID(t *testing.T) {
	sub, key, alg, err := ParseKeyID("p2p-trading-sandbox1.com|76EU8aUq|ed25519")
	if err != nil {
		t.Fatalf("ParseKeyID() error: %v", err)
	}
	if sub != "p2p-trading-sandbox1.com" || key != "76EU8aUq" || alg != "ed25519" {
		t.Errorf("unexpected result: %s, %s, %s", sub, key, alg)
	}

	_, _, _, err = ParseKeyID("bad-format")
	if err == nil {
		t.Fatal("expected error for bad format")
	}
}

File: DEG-main/devkits/p2p-trading-ies-wave1/beckn-signing-kit/nodejs/signer.test.js
'use strict';

const { describe, it } = require('node:test');
const assert = require('node:assert/strict');
const crypto = require('crypto');
const { PayloadSigner, verify, verifyAt, parseKeyId } = require('./index');

// Test keys matching the config in local-p2p-bap.yaml (sandbox1).
const TEST_SUBSCRIBER_ID = 'p2p-trading-sandbox1.com';
const TEST_KEY_ID = '76EU8aUqHouww7gawT6EibH4bseMCumyDv3sgyXSKENGk8NDcdVwmQ';
const TEST_PRIVATE_KEY = 'Pc6dkYo5LeP0LkwvZXVRV9pcbeh8jDdtdHWymID5cjw=';
const TEST_PUBLIC_KEY = 'KVYEWkQB2WwnttVMWfy7KrnqiD51ZDvi8vfCac2IwRE=';

// Sample beckn payload (trimmed confirm request).
const samplePayload = Buffer.from(JSON.stringify({
  context: {
    version: '2.0.0',
    action: 'confirm',
    timestamp: '2024-10-04T10:25:00Z',
    message_id: 'msg-confirm-001',
    transaction_id: 'txn-energy-001',
    bap_id: 'bap.energy-consumer.com',
    bap_uri: 'https://bap.energy-consumer.com',
    bpp_id: 'bpp.energy-provider.com',
    bpp_uri: 'https://bpp.energy-provider.com',
    domain: 'beckn.one:deg:p2p-trading:2.0.0',
  },
  message: {
    order: {
      '@type': 'beckn:Order',
      'beckn:orderStatus': 'CREATED',
      'beckn:seller': 'provider-solar-farm-001',
    },
  },
}));

function makeSigner(overrides) {
  return new PayloadSigner({
    subscriberId: TEST_SUBSCRIBER_ID,
    uniqueKeyId: TEST_KEY_ID,
    signingPrivateKey: TEST_PRIVATE_KEY,
    ...overrides,
  });
}

describe('PayloadSigner', () => {
  describe('constructor validation', () => {
    it('rejects missing subscriberId', () => {
      assert.throws(
        () => new PayloadSigner({ uniqueKeyId: 'k', signingPrivateKey: TEST_PRIVATE_KEY }),
        /subscriberId/,
      );
    });

    it('rejects missing uniqueKeyId', () => {
      assert.throws(
        () => new PayloadSigner({ subscriberId: 's', signingPrivateKey: TEST_PRIVATE_KEY }),
        /uniqueKeyId/,
      );
    });

    it('rejects missing signingPrivateKey', () => {
      assert.throws(
        () => new PayloadSigner({ subscriberId: 's', uniqueKeyId: 'k' }),
        /signingPrivateKey/,
      );
    });

    it('rejects invalid base64', () => {
      assert.throws(
        () => new PayloadSigner({ subscriberId: 's', uniqueKeyId: 'k', signingPrivateKey: 'not-base64!!!' }),
        /invalid base64/,
      );
    });

    it('rejects wrong key size', () => {
      const shortKey = Buffer.from('short').toString('base64');
      assert.throws(
        () => new PayloadSigner({ subscriberId: 's', uniqueKeyId: 'k', signingPrivateKey: shortKey }),
        /must be 32 bytes/,
      );
    });
  });

  describe('signPayload', () => {
    it('produces a valid Authorization header', () => {
      const signer = makeSigner();
      const authHeader = signer.signPayload(samplePayload);

      assert.ok(authHeader.startsWith('Signature '), 'should start with "Signature "');
      assert.ok(
        authHeader.includes(`${TEST_SUBSCRIBER_ID}|${TEST_KEY_ID}|ed25519`),
        'should contain keyId',
      );
      assert.ok(authHeader.includes('algorithm="ed25519"'), 'should contain algorithm');
    });
  });

  describe('signPayloadDetailed', () => {
    it('returns correct expiry window', () => {
      const signer = makeSigner();
      const result = signer.signPayloadDetailed(samplePayload);

      assert.equal(result.expiresAt - result.createdAt, 300, 'default expiry should be 300s');
      assert.ok(result.signature, 'signature should be non-empty');
      assert.ok(result.authorizationHeader, 'authorizationHeader should be non-empty');
    });

    it('respects custom expiry', () => {
      const signer = makeSigner({ expirySeconds: 60 });
      const result = signer.signPayloadDetailed(samplePayload);

      assert.equal(result.expiresAt - result.createdAt, 60);
    });
  });

  describe('accepts string body', () => {
    it('signs string payload identically to buffer', () => {
      const signer = makeSigner();
      const payloadStr = samplePayload.toString();

      // Both should produce valid signatures (can't compare directly due to timestamp)
      const authHeader = signer.signPayload(payloadStr);
      assert.ok(authHeader.startsWith('Signature '));
    });
  });
});

describe('sign and verify round-trip', () => {
  it('verifies a freshly signed payload', () => {
    const signer = makeSigner();
    const authHeader = signer.signPayload(samplePayload);

    // Should not throw
    verify(samplePayload, authHeader, TEST_PUBLIC_KEY);
  });

  it('rejects a tampered payload', () => {
    const signer = makeSigner();
    const authHeader = signer.signPayload(samplePayload);

    const tampered = Buffer.from(samplePayload);
    tampered[10] = tampered[10] ^ 0xff;

    assert.throws(
      () => verify(tampered, authHeader, TEST_PUBLIC_KEY),
      /verification failed/,
    );
  });

  it('rejects an expired signature', () => {
    const signer = makeSigner({ expirySeconds: 1 });
    const pastTime = Math.floor(Date.now() / 1000) - 600; // 10 minutes ago
    const result = signer._signPayloadAt(samplePayload, pastTime);

    assert.throws(
      () => verify(samplePayload, result.authorizationHeader, TEST_PUBLIC_KEY),
      /expired/,
    );
  });

  it('rejects a wrong public key', () => {
    const signer = makeSigner();
    const authHeader = signer.signPayload(samplePayload);

    // Generate a different key pair
    const { publicKey } = crypto.generateKeyPairSync('ed25519');
    const rawPub = publicKey.export({ format: 'der', type: 'spki' });
    // Extract the raw 32-byte key from SPKI DER (last 32 bytes)
    const wrongPubB64 = rawPub.subarray(rawPub.length - 32).toString('base64');

    assert.throws(
      () => verify(samplePayload, authHeader, wrongPubB64),
      /verification failed/,
    );
  });
});

describe('verifyAt', () => {
  it('rejects signature not yet valid', () => {
    const signer = makeSigner();
    const futureTime = Math.floor(Date.now() / 1000) + 3600;
    const result = signer._signPayloadAt(samplePayload, futureTime);

    // Verify at current time (before created)
    const nowUnix = Math.floor(Date.now() / 1000);
    assert.throws(
      () => verifyAt(samplePayload, result.authorizationHeader, TEST_PUBLIC_KEY, nowUnix),
      /not yet valid/,
    );
  });
});

describe('parseKeyId', () => {
  it('parses a valid keyId', () => {
    const result = parseKeyId('p2p-trading-sandbox1.com|76EU8aUq|ed25519');
    assert.deepEqual(result, {
      subscriberId: 'p2p-trading-sandbox1.com',
      uniqueKeyId: '76EU8aUq',
      algorithm: 'ed25519',
    });
  });

  it('rejects invalid format', () => {
    assert.throws(() => parseKeyId('bad-format'), /invalid keyId format/);
  });
});
